DRIVERS-3329: Configurable DNS domain validation for SRV records (#1950)
mongodb/specifications: source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.md

Reviewer: There's no security-considerations text for this new option. The Design Rationale section is a single pre-existing sentence and doesn't mention `srvAllowedHostsSuffix`. I think we could benefit from a subsection covering: why the option exists; that the PSL bounds the blast radius but doesn't make a configuration safe; guidance to use the narrowest suffix that work. PyMongo already has good material for most of this in its MongoClient docstring warning. Also helps future spec readers.

Author: added in [d2731ce](https://github.com/mongodb/specifications/pull/1950/commits/d2731ceb4b7619cb6a0ab727337afe9cbc3c5ff7)
honestly i feel a bit off about my wording though? so lmk if you have any suggestions on the wording here!

Reviewer: I think the section reads well! I wouldn't change anything. One thing I would suggest adding though is info around the alternative @damieng raised - the driver provided hook. Something like 

Why not a validation callback?
An alternative considered was a driver-provided hook....bla.....bla.... bla

`{options}` refers to the optional elements from the [Connection String](../connection-string/connection-string-spec.md)
specification following the `Host Information`. This includes the `Auth database` and `Connection Options`.

For the purposes of this document, `{hostname}` will be divided using the following terminology. If

> `srvAllowedHostsSuffix` has been configured, then that will act as the `{domainname}`. Otherwise, if an SRV `{hostname}`

Reviewer: Question for the rationale rather than a change request: nothing requires `{hostname}` itself to end with the suffix, so `mongodb+srv://cluster.example.org/?srvAllowedHostsSuffix=bank.com` is legal.

I initially thought that should be forbidden, but I don't think it should — it would break the vanity-CNAME shape, where the SRV name is a corporate alias and the hosts live in a provider domain, which seems like one of the cases this option might be used for. So I'd just like the rationale to say that decoupling is deliberate and why, since it's the point where the trust anchor stops being tied to the name the user typed.

Author: that is correct, I've added this to the rationale in [d2731ce](https://github.com/mongodb/specifications/pull/1950/commits/d2731ceb4b7619cb6a0ab727337afe9cbc3c5ff7)

### MongoClient Configuration

#### srvAllowedHostsSuffix

> This option is used to validate hosts. If present, its value MUST be treated as the domain for DNS validation. For

Reviewer 2: Optional: Consider linking to the [Querying DNS](https://github.com/mongodb/specifications/blob/fea49b840546aa81f42d4b4578820eebfac0d4b9/source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.md?plain=1#L114) section, which is where the DNS validation is described.

Author: added! thanks!

> This option is used to validate hosts. If present, its value MUST be treated as the domain for DNS validation. For

Reviewer 2: Use the well defined value `{domainname}` instead of "domain" to clarify how `srvAllowedHostsSuffix` should be used during DNS hostname validation.

```suggestion
This option is used to validate hosts. If present, its value MUST be treated as the `{domainname}` for DNS validation. For
```

example, `srvAllowedHostsSuffix=.mydomain.net`. If the value does not begin with a `.`, for example,
`srvAllowedHostsSuffix=mydomain.net`, the `.` MUST be automatically prepended prior to validation. If this option is not

> present, the domain MUST be inferred from the hostname. This option MUST only be configurable at the level of a

Reviewer 2: Use the well defined values `{domainname}` and `{hostname}`

Optional: Consider adding a link to "Connection String Format", which is where the logic for deriving `{domainname}` from `{hostname}` is described.

```suggestion
present, the`{domainname}` MUST be inferred from the `{hostname}` (as described in [Connection String Format](#connection-string-format)). This option MUST only be configurable at the level of a
```

> `MongoClient`.

Reviewer 3: Just to clarify, if a host is invalid eg: `..example.com` should this error or just return an empty list of hosts?

Author: oh good question, my current implementation has it _eventually_ erroring (with a configuration error) when the driver goes to validate the SRV hosts -- i didn't add a new "check valid host" type of function. (though this probably wouldn't be too hard with some regex? if you think it'd be beneficial to add something like this.) My initial thought process was this new parameter is just a user configurable way to denote what the domain should be (as opposed to the previous, now default, logic) -- they should know what they're doing if they're using this param and if they happened to make a typo / silly mistake, the existing "invalid SRV host" error message that occurs when there is an invalid host given the previous logic would still apply.

Reviewer 3: I think the eventually error part will cover this, so no need for a regex. I can never understand regexes and someone always seems to find a usecase where the regex fails where it shouldn't!

### MongoClient Configuration

#### srvAllowedHostsSuffix

This option is used to validate hosts. If present, its value MUST be treated as the `{domainname}` for
[DNS validation](#querying-dns). For example, `srvAllowedHostsSuffix=.mydomain.net`. If the value does not begin with a
`.`, for example, `srvAllowedHostsSuffix=mydomain.net`, the `.` MUST be automatically prepended prior to validation. If

> this option is not present, the`{domainname}` MUST be inferred from the `{hostname}` (as described in

Reviewer 4: Missing space

### MongoClient Configuration

#### srvAllowedHostsSuffix

This option is used to validate hosts. If present, its value MUST be treated as the `{domainname}` for
[DNS validation](#querying-dns) and

> [SRV polling](../polling-srv-records-for-mongos-discovery/polling-srv-records-for-mongos-discovery.md). For example,

Reviewer: The option description now says it applies to SRV polling but that spec document isn't touched by this PR and still carries its own normative sentence: `"A driver MUST verify that the host names returned through SRV records have the same parent {domainname}."` 

More subtly, the two documents now use `{domainname}` to mean different things — in the polling spec it's the `{hostname}.{domainname}` split; here it's the entire configured suffix when the option is set. WE should update SRV polling spec accordingly as well.

Author: oo thanks for pointing that out! fixed in [d2731ce](https://github.com/mongodb/specifications/pull/1950/commits/d2731ceb4b7619cb6a0ab727337afe9cbc3c5ff7)

`srvAllowedHostsSuffix=.mydomain.net`. Drivers MUST apply the following normalization and validation to the value before
use:

- If the value does not begin with a `.`, a `.` MUST be automatically prepended. For example,
    `srvAllowedHostsSuffix=mydomain.net` is treated as `.mydomain.net`.

> - The value MUST NOT be a public suffix, per the algorithm in (../public-suffix-list/public-suffix-list.md)

Reviewer: Is this meant to be rendered like this or is it meant to be a clickable link to that path?

> - The value MUST be normalized to lowercase using ASCII case folding before comparison.

Reviewer: I think these bullets need to be an ordered sequence rather than a set, and two things are missing.
- Trailing dots. `srvAllowedHostsSuffix-trailing-dot` requires `.build.10gen.cc.` to succeed, but nothing here mandates stripping it — the trailing-dot rule added under "Querying DNS" is scoped to returned hostnames. The test currently requires unspecified behavior.
- `IDN` (related to Arthurs questions).  the PSL side is covered — [upstream's algorithm](https://github.com/publicsuffix/list/wiki/Format#formal-algorithm) requires Punycode canonicalization before comparison. But the suffix-vs-hostname comparison isn't a PSL operation, and these bullets specify only ASCII case folding. A Unicode value silently never matches an A-label host, so the user gets "hosts don't match your suffix" rather than "your suffix is malformed." Converting also forecloses homographs by ruling out Unicode case folding or NFC.
-  Also, the middle bullet is missing the [...] half of its link (I assuming it is meant to be link to public suffix list section added by PR 1972).
```suggestion
`srvAllowedHostsSuffix=.mydomain.net`. Drivers MUST apply the following normalization and validation to the value, in this order:
1. Any trailing `.` MUST be stripped. For example, `srvAllowedHostsSuffix=.mydomain.net.` is treated as `.mydomain.net`.
2. The value MUST be converted to its A-label (Punycode) form, so that it is comparable against the A-label hostnames returned by DNS.
3. The value MUST be normalized to lowercase using ASCII case folding.
4. The resulting value, with any leading `.` removed, MUST NOT be a public suffix, per the algorithm in [Public Suffix List](../public-suffix-list/public-suffix-list.md).
5. If the value does not begin with a `.`, a `.` MUST be prepended. For example, `srvAllowedHostsSuffix=mydomain.net` is treated as `.mydomain.net`.
```

Step 4 removing the leading dot matters — upstream's algorithm forbids empty labels, so `.com` isn't a valid input to the lookup.

### MongoClient Configuration

#### srvAllowedHostsSuffix

This option is used to validate hosts. If present, its value MUST be treated as the `{domainname}` for
[DNS validation](#querying-dns) and
[SRV polling](../polling-srv-records-for-mongos-discovery/polling-srv-records-for-mongos-discovery.md). For example,
`srvAllowedHostsSuffix=.mydomain.net`. Drivers MUST apply the following normalization and validation to the value, in
this order:

1. Any trailing `.` MUST be stripped. For example, `srvAllowedHostsSuffix=.mydomain.net.` is treated as `.mydomain.net`.
2. The value MUST be converted to its A-label (Punycode) form, so that it is comparable against the A-label hostnames
    returned by DNS.
3. The value MUST be normalized to lowercase using ASCII case folding.
4. The resulting value, with any leading `.` removed, MUST NOT be a public suffix, per the algorithm in
    [Public Suffix List](../public-suffix-list/public-suffix-list.md).
5. If the value does not begin with a `.`, a `.` MUST be prepended. For example, `srvAllowedHostsSuffix=mydomain.net` is

>     treated as `.mydomain.net`.

Reviewer: Small ordering fix I missed from my previous review comments: step 2 runs while the value can still begin with ".", which is an empty label. IDNA rejects those, so the standard conversion function throws on the example value in this very section — str.encode('idna') and the idna package both raise on `.mydomain.net`, as does .NET's `IdnMapping.GetAscii`. Stripping the leading dot alongside the trailing one fixes it and lets steps 4 and 5 drop their conditionals. No behavior change, no test changes.
```suggestion
1. Any leading or trailing `.` MUST be stripped. For example, `srvAllowedHostsSuffix=.mydomain.net.` is treated as `mydomain.net`.
2. The value MUST be converted to its A-label (Punycode) form, so that it is comparable against the A-label hostnames returned by DNS.
3. The value MUST be normalized to lowercase using ASCII case folding.
4. The resulting value MUST NOT be a public suffix, per the algorithm in [Public Suffix List](../public-suffix-list/public-suffix-list.md).
5. A `.` MUST be prepended. For example, `srvAllowedHostsSuffix=mydomain.net` is treated as `.mydomain.net`.
```

indicating that the URI could not be used to find hostnames. The error SHALL include the reason why they could not be
found.

Before validating returned hostnames, drivers MUST normalize them as follows:

- Trailing dots MUST be stripped (e.g. `host.example.com.` becomes `host.example.com`).
- Hostnames MUST be normalized to lowercase using ASCII case folding.

A driver MUST verify that the host names returned through SRV records share the original SRV's `{domainname}`. In
addition, SRV records with fewer than three `.` separated parts, the returned hostname MUST have at least one more

> domain level than the SRV record hostname. Drivers MUST raise an error and MUST NOT initiate a connection to any

Reviewer: I think this part "In addition, SRV records with fewer than three . separated parts…" doesn't apply when  `srvAllowedHostsSuffix` is set. The clause needs `{hostname}` to have ≤2 labels, the suffix must have ≥2 labels (a 1-label suffix is already a public suffix under the * fallback), and the mandatory leading dot forces returned hosts to ≥3 labels. So the "at least one more domain level" property holds by construction. I suggest updating this part of the spec.

Author: good catch! done in [d2731ce](https://github.com/mongodb/specifications/pull/1950/commits/d2731ceb4b7619cb6a0ab727337afe9cbc3c5ff7)

indicating that the URI could not be used to find hostnames. The error SHALL include the reason why they could not be
found.

Before validating returned hostnames, drivers MUST normalize them as follows:

- Trailing dots MUST be stripped (e.g. `host.example.com.` becomes `host.example.com`).

> - Hostnames MUST be normalized to lowercase using ASCII case folding.

Reviewer: Because this addition sits in a PR titled after the new option, it could be read as a detail of that option — I think it earns its own entry in this document's ChangeLog so other drivers audit their comparison instead of skipping it:
 - 2026-08-24: Specify that host names returned through SRV records are normalized -- trailing dot stripped, ASCII lowercase -- before validation.

Also, nothing in the test suite can catch a driver that skips the fold — every `build.10gen.cc` record returns lowercase targets, so it's unobservable, and `srvAllowedHostsSuffix-case-insensitive` covers the suffix side only. I think we need a fixture with a mixed-case SRV target, which is test-infra work but worth doing in this PR. Flagging it because a driver can be compliant on the seedlist path and not on the polling path.

Author: makes sense, added changelog suggestion 

oh that's a good catch -- filed a devprod ticket ([DEVPROD-42090](https://jira.mongodb.org/browse/DEVPROD-42090)) I hope that's the right thing to do? (and hopefully it doesn't take long for them to get back to us?)
